server/analyzer/replace_arithmetic_expressions.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2024 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package analyzer
+
+import (
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/analyzer"
+	gms_expression "github.com/dolthub/go-mysql-server/sql/expression"
+	"github.com/dolthub/go-mysql-server/sql/plan"
+	"github.com/dolthub/go-mysql-server/sql/transform"
+
+	"github.com/dolthub/doltgresql/server/expression"
+	"github.com/dolthub/doltgresql/server/functions/framework"
+)
+
+func ReplaceArithmeticExpressions(ctx *sql.Context, a *analyzer.Analyzer, node sql.Node, scope *plan.Scope, selector analyzer.RuleSelector, qFlags *sql.QueryFlags) (sql.Node, transform.TreeIdentity, error) {
+	return transform.NodeExprsWithOpaque(node, func(e sql.Expression) (sql.Expression, transform.TreeIdentity, error) {
+		switch e := e.(type) {
+		case *gms_expression.Arithmetic:
+			switch e.Operator() {
+			case "+":
+				expr, err := expression.NewBinaryOperator(framework.Operator_BinaryPlus).WithResolvedChildren(childrenAsAnySlice(e))
```

Comment:
Seems **way** better to replace this with a call to `WithChildren`. Currently `WithChildren` makes an assumption that we've already found the operator function, but we could change it to:
```go
func (b *BinaryOperator) WithChildren(children ...sql.Expression) (sql.Expression, error) {
	if len(children) != 2 {
		return nil, sql.ErrInvalidChildrenNumber.New(b, len(children), 2)
	}
	if b.compiledFunc != nil {
		compiledFunc, err := b.compiledFunc.WithChildren(children...)
		if err != nil {
			return nil, err
		}
		return &BinaryOperator{
			operator:     b.operator,
			compiledFunc: compiledFunc.(framework.Function),
		}, nil
	} else {
		binOp, err := b.WithResolvedChildren([]any{children[0], children[1]})
		if err != nil {
			return nil, err
		}
		return binOp.(sql.Expression), nil
	}
}
```
Using `WithResolvedChildren` outside of the GMS-only expectation adds a potential logical dependency that we probably shouldn't make a precedent for. If `WithChildren` calls it internally, then its usage is more of an implementation detail than a sort of API contract.

We could also make this `WithChildren` change to `unary_operator` since they're very similar expressions.

Reply:
Done